android.os.TransactionTooLargeException

java.lang.RuntimeException: 
  at android.app.ActivityThread$StopInfo.run (ActivityThread.java:3797)
  at android.os.Handler.handleCallback (Handler.java:751)
  at android.os.Handler.dispatchMessage (Handler.java:95)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6171)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:891)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:781)

Caused by: android.os.TransactionTooLargeException: 
  at android.os.BinderProxy.transactNative (Native Method)
  at android.os.BinderProxy.transact (Binder.java:615)
  at android.app.ActivityManagerProxy.activityStopped (ActivityManagerNative.java:3650)
  at android.app.ActivityThread$StopInfo.run (ActivityThread.java:3789)

Or

java.lang.RuntimeException:  
  at android.app.ActivityThread$StopInfo.run (ActivityThread.java:3825)
  at android.os.Handler.handleCallback (Handler.java:751)
  at android.os.Handler.dispatchMessage (Handler.java:95)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6236)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:891)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:781)
  at de.robv.android.xposed.XposedBridge.main (XposedBridge.java:103)

Caused by: android.os.TransactionTooLargeException: 
  at android.os.BinderProxy.transactNative (Native Method)
  at android.os.BinderProxy.transact (Binder.java:615)
  at android.app.ActivityManagerProxy.activityStopped (ActivityManagerNative.java:3650)
  at android.app.ActivityThread$StopInfo.run (ActivityThread.java:3817)

<img width="716" alt="screen shot 2017-10-15 at 7 39 07 pm" src="https://user-images.githubusercontent.com/6244235/31584398-46abdc74-b1e0-11e7-887e-068da3dbbf4d.png">
<img width="979" alt="screen shot 2017-10-15 at 7 39 21 pm" src="https://user-images.githubusercontent.com/6244235/31584399-4968be3c-b1e0-11e7-8b61-b1c7cf1cf759.png">

--- a/app/src/main/java/org/mozilla/focus/fragment/WebFragment.java
+++ b/app/src/main/java/org/mozilla/focus/fragment/WebFragment.java
@@ -25,6 +25,9 @@ import org.mozilla.focus.webkit.WebkitView;
  * Base implementation for fragments that use an IWebView instance. Based on Android's WebViewFragment.
  */
 public abstract class WebFragment extends LocaleAwareFragment {
+
+    private static final int BUNDLE_MAX_SIZE = 300 * 1000; // 300K
+
     private IWebView webView;
     // webView is not available after onDestroyView, but we need webView reference in callback
     // onSaveInstanceState. However the callback might be invoked at anytime before onDestroy
@@ -110,6 +113,18 @@ public abstract class WebFragment extends LocaleAwareFragment {
     @Override
     public void onSaveInstanceState(Bundle outState) {
         webView.onSaveInstanceState(outState);
+
+        // Workaround for #1107 TransactionTooLargeException
+        // since Android N, system throws a exception rather than just a warning(then drop bundle)
+        // To set a threshold for dropping WebView state manually
+        // refer: https://issuetracker.google.com/issues/37103380
+        final String key = "WEBVIEW_CHROMIUM_STATE";
+        if (outState.containsKey(key)) {
+            final int size = outState.getByteArray(key).length;
+            if (size > BUNDLE_MAX_SIZE) {
+                outState.remove(key);
+            }
+        }
 
         super.onSaveInstanceState(outState);
     }

